File: 04-implementation/backend/app/models/vocab.py
from sqlalchemy import String, Integer, Text, Float, ARRAY
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class SkillCluster(Base):
    """Canonical skill in the emergent vocabulary."""
    __tablename__ = "skill_clusters"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    label: Mapped[str] = mapped_column(String(512), nullable=False, unique=True)
    version: Mapped[int] = mapped_column(Integer, default=1)


class SkillToken(Base):
    """Raw extracted skill term mapped to a cluster."""
    __tablename__ = "skill_tokens"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    raw_text: Mapped[str] = mapped_column(String(512), nullable=False)
    cluster_id: Mapped[int] = mapped_column(Integer, nullable=True)
    embedding_json: Mapped[str] = mapped_column(Text, nullable=True)  # JSON float array

File: 04-implementation/backend-rust/shared/src/models/vocab.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct SkillCluster {
    pub id: i32,
    pub label: String,
    pub version: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct SkillToken {
    pub id: i32,
    pub raw_text: String,
    pub cluster_id: Option<i32>,
    pub embedding_json: Option<String>,
}
